microsetta_admin/metadata_transforms.py

```diff
@@ -0,0 +1,236 @@
+        series.name = cls.COLUMN_NAME
+        return series
+
+
+def _normalizer(df, focus_col, units_col, units_value, factor):
+    # get our columns
+    focus = pd.to_numeric(df[focus_col], errors='coerce')
+    units = df[units_col]
+
+    # operate on a copy as to retain non-unit focus values (e.g., values
+    # already expressed as centimeters
+    result = focus.copy()
+
+    # anything negative is weird so kill it
+    result[result < 0] = None
+
+    # figure out what positions are safe to operate on
+    not_null_map = Transformer.not_null_map(result.isnull(),
+                                            units.isnull())
+
+    # take entries like where either focus or units are null and kill them
+    result.loc[not_null_map[~not_null_map].index] = None
+
+    # reduce to only those safe to operate on
+    focus_not_null = result[not_null_map]
+    units_not_null = units[not_null_map]
+    focus_adj = focus_not_null.loc[units_not_null == units_value]
+
+    # adjust the indices that need adjustment
+    result.loc[focus_adj.index] = focus_adj * factor
+
+    return result.fillna(MISSING_VALUE)
+
+
+class _Normalize(Transformer):
+    FOCUS_COL = None
+    FOCUS_UNITS = None
+    UNITS_COL = None
+    FACTOR = None
+
+    @classmethod
+    def _transform(cls, df):
+        return _normalizer(df, cls.FOCUS_COL, cls.UNITS_COL, cls.FOCUS_UNITS,
+                           cls.FACTOR)
+
+
+class NormalizeHeight(_Normalize):
+    REQUIRED_COLUMNS = frozenset(['height_units', 'height_cm'])
+    COLUMN_NAME = 'height_cm'
+    UPDATE_EXISTING = ('height_units', 'centimeters')
+    FOCUS_COL = 'height_cm'
+    UNITS_COL = 'height_units'
+    FOCUS_UNITS = 'inches'
+    FACTOR = 2.54
+
+
+class NormalizeWeight(_Normalize):
+    REQUIRED_COLUMNS = frozenset(['weight_units', 'weight_cm'])
+    COLUMN_NAME = 'weight_kg'
+    UPDATE_EXISTING = ('weight_units', 'kilograms')
+    FOCUS_COL = 'weight_kg'
+    FOCUS_UNITS = 'pounds'
+    UNITS_COL = 'weight_units'
+    FACTOR = (1 / 2.20462)
+
+
+# transforms are order dependent as some entries (e.g., BMICat) depend
+# on the presence of a BMI column
+HUMAN_TRANSFORMS = (AgeYears, AgeCat, NormalizeWeight, NormalizeHeight,
+                    BMI, BMICat, AlcoholConsumption)
+
+
+def apply_category_specific_transforms(df, transforms):
+    for transform in transforms:
+        # new columns are potentially added on each transform so recollect
+        columns = set(df.columns)
+
+        if transform.REQUIRED_COLUMNS.issubset(columns):
+            # note: not using df.apply here as casts are needed on a
+            # case-by-case basis, and pandas is much more efficient
+            # casting whole columns
+            subset = df[transform.REQUIRED_COLUMNS]
+            series = transform.apply(subset)
+            df[transform.COLUMN_NAME] = series.astype(str)
+
+            # update a an existing column if the transform needs to
+            if transform.UPDATE_EXISTING is not None:
```

Comment:
The generic name "UPDATE_EXISTING" made me expect a rather generalized functionality for updating however many existing columns I wanted.  Since this is actually (and reasonably) a functionality that is really aimed only at updating a single units column to a single, pre-defined value, can we rename to something like "EXISTING_UNITS_COL_UPDATE" or something?